### PR TITLE
feat(role-permissions): request the new permissions format

### PR DESCRIPTION
**Pull request by adriguy**
## Pull Request checklist:

- [x] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [ ] Wonder if you can improve the existing code

### CODE REVIEW

File: src/services/permissions-checker.js [hunk 1]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
...
+  }
+
+  static transformPermissionsFromOldToNewFormat(permissions) {
+    Object.keys(permissions).forEach((modelName) => {
+      const { collection } = permissions[modelName];
+      permissions[modelName].collection = {
+        browseEnabled: collection.list || collection.searchToEdit,
+        readEnabled: collection.show,
+        editEnabled: collection.create,
+        addEnabled: collection.update,
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:57]: Ahahah you switch the value 😂

File: src/services/permissions-checker.js [hunk 2]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
...
+      && PermissionsChecker.permissions.renderings[this.renderingId].data[collectionName]) {
+      return PermissionsChecker.permissions.renderings[this.renderingId].data[collectionName].scope;
+    }
+    return null;
+  }
+
+  static transformPermissionsFromOldToNewFormat(permissions) {
+    Object.keys(permissions).forEach((modelName) => {
+      const { collection } = permissions[modelName];
+      permissions[modelName].collection = {
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:53]: define a new object instead of reusing the same object.
For 2 reasons:
- Your function is modifying a parameter which is really bad. We don't want that.
- Your permissions will actually contain the new AND the old format 😅

File: src/services/permissions-checker.js [hunk 3]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
...
+        browseEnabled: collection.list || collection.searchToEdit,
+        readEnabled: collection.show,
+        editEnabled: collection.create,
+        addEnabled: collection.update,
+        deleteEnabled: collection.delete,
+        exportEnabled: collection.export,
+      };
+
+      if (permissions[modelName].actions) {
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:62]: You are using `permissions[modelName]` like a thousand times 😅  you should define a variable 😉

File: src/services/permissions-checker.js [hunk 4]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
...
+
+  static getLastRetrieveTime(renderingId, permissionName) {
+    // In the case of rolesACL format and browseEnabled permission, the last retrieve to be taken
+    // into account is the one stored by rendering (because of the scope information).
+    if (PermissionsChecker.isrolesACLActivated && permissionName !== 'browseEnabled') {
+      return PermissionsChecker.permissions.collections.lastRetrieve;
+    }
+    return PermissionsChecker.permissions.renderings[renderingId]
+      ? PermissionsChecker.permissions.renderings[renderingId].lastRetrieve
+      : null;
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:113]: This is due to the fact that you use 2 different formats when storing your data. You should not change this function

File: src/services/permissions-checker.js [hunk 5]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
...
   static resetExpiration(renderingId) {
-    const permissions = PermissionsChecker.getPermissions(renderingId);
-    if (permissions) {
-      permissions.lastRetrieve = null;
+    if (PermissionsChecker.isrolesACLActivated && PermissionsChecker.permissions.collection) {
+      PermissionsChecker.permissions.collection.lastRetrieve = null;
+    }
+
+    if (PermissionsChecker.permissions.renderings[renderingId]) {
+      PermissionsChecker.permissions.renderings[renderingId].lastRetrieve = null;
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:122]: Nor this one !

File: src/services/permissions-checker.js [hunk 6]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
...
     }
   }
 
+  static _isPermissionAllowed(permissionValue, userId) {
+    return Array.isArray(permissionValue)
+      ? permissionValue.includes(parseInt(userId, 10))
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:128]:
```suggestion
      ? permissionValue.includes(Number.parseInt(userId, 10))
```

File: src/services/permissions-checker.js [hunk 7]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
...
-    const computedConditionFilters = _.clone(collectionListScope.filter);
+  static _computeConditionFiltersFromScope(userId, scope) {
+    const computedConditionFilters = _.clone(scope.filter);
     computedConditionFilters.conditions.forEach((condition) => {
       if (condition.value
         && `${condition.value}`.startsWith('$')
-        && collectionListScope.dynamicScopesValues.users[userId]) {
-        condition.value = collectionListScope
+        && scope.dynamicScopesValues.users[userId]) {
+        condition.value = scope
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:156]: This looks like a refacto and should not be part of this PR

File: src/services/permissions-checker.js [hunk 8]
```diff
@@ -101,89 +187,103 @@ class PermissionsChecker {
...
-        && expectedConditionFilters.conditions.length === 1;
+  static async _isCollectionBrowseAllowed(collectionPermissions, permissionInfos, scope) {
+    const { browseEnabled } = collectionPermissions.collection;
+    const { userId } = permissionInfos;
 
-      const isSameScope = !!scopeFound
-        && scopeFound.aggregator === expectedConditionFilters.aggregator
-        && !!scopeFound.conditions
-        && scopeFound.conditions.length === expectedConditionFilters.conditions.length;
+    if ((Array.isArray(browseEnabled) && !browseEnabled.includes(parseInt(userId, 10)))
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:229]: You've created a function that does exactly that 😅 `_isPermissionAllowed` 🤔

File: test/services/permissions.test.js [hunk 9]
```diff
@@ -9,154 +9,164 @@ describe('services > permissions', () => {
...
+                  },
                 },
-              },
-            });
+              });
 
-          await new PermissionsChecker('envSecret', 1).checkPermissions('Users', 'list')
-            .then(() => { expect(true).toStrictEqual(true); });
+            await new PermissionsChecker('envSecret', 1).checkPermissions('Users', 'browseEnabled', { userId: 1 })
+              .then(() => { expect(true).toStrictEqual(true); });
```
Comment (VincentMolinie) [inline test/services/permissions.test.js:35]: You are using `await` so get rid of the `then` it will be much easier to read

File: test/services/permissions.test.js [hunk 10]
```diff
@@ -9,154 +9,164 @@ describe('services > permissions', () => {
...
+                      list: true,
+                    },
+                  },
+                },
+              });
 
-        await expect(new PermissionsChecker('envSecret', 1).checkPermissions('Users', 'list'))
-          .rejects.toThrow("'list' access forbidden on Users");
+            await new PermissionsChecker('envSecret', 2).checkPermissions('Users', 'browseEnabled', { userId: 1 })
+              .then(() => { expect(true).toStrictEqual(true); });
```
Comment (VincentMolinie) [inline test/services/permissions.test.js:131]: Looking at your test you should use `.not.toThrow()` which is way more explicit than `expect(true).toStrictEqual(true)` 😅

File: test/services/permissions.test.js [hunk 11]
```diff
@@ -404,257 +559,1036 @@ describe('services > permissions', () => {
...
+                },
+              });
+
+              const smartActionParameters = {
+                actionName: 'save',
+                userId: '1',
+              };
+
+              const result = await new PermissionsChecker('envSecret', 1).checkPermissions('Users', 'actions', smartActionParameters);
+              expect(result).toBeUndefined();
```
Comment (VincentMolinie) [inline test/services/permissions.test.js:1138]: What is that test 😅 ? Are you testing that it does not throw ?

**Review by VincentMolinie — CHANGES_REQUESTED** [review]
Nice that you implemented all those tests but it's way to hard to follow 😅 ... It would have been better to just implement Unit test so it's easier to read and understand ^^

File: src/services/permissions-checker.js [hunk 12]
```diff
@@ -16,58 +16,144 @@ class PermissionsChecker {
 
   static expirationInSeconds = EXPIRATION_IN_SECONDS;
 
-  // This permissionsPerRendering is the cache, shared by all instances of PermissionsChecker.
-  static permissionsPerRendering = {};
+  static isrolesACLActivated;
```
[the remark this answers is not in the document]
Comment (adriguy) [reply src/services/permissions-checker.js:19]: Good catch!! 🙏

File: src/services/permissions-checker.js (hunk 2 above)
Comment (adriguy) [reply src/services/permissions-checker.js:53]: Not sure why it will contain the old format 🤔 
But sure it would be better to user a new object here! 🙏

File: src/services/permissions-checker.js (hunk 7 above)
Comment (adriguy) [reply src/services/permissions-checker.js:156]: Yes sorry, hopefully you don't mind!

File: src/services/permissions-checker.js (hunk 8 above)
Comment (adriguy) [reply src/services/permissions-checker.js:229]: 😄  OMG that's true thanks!!

File: test/services/permissions.test.js (hunk 9 above)
Comment (adriguy) [reply test/services/permissions.test.js:35]: This is old code but you're right

File: test/services/permissions.test.js (hunk 11 above)
Comment (adriguy) [reply test/services/permissions.test.js:1138]: Yes sorry I copied it from the existing tests... I guess it is just checking that the permission is allowed, so that it doesn't throw indeed!
I'll change them all.

File: src/services/permissions-checker.js (hunk 1 above)
Comment (adriguy) [reply src/services/permissions-checker.js:57]: OMG thanks!

File: src/services/permissions-cache.js [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    if (PermissionsCache.isRolesACLActivated) {
+      return collectionsPermissions ? collectionsPermissions.data : null;
+    }
+    return renderingPermissions ? renderingPermissions.data : null;
+  }
+
+  static getScopePermissions(renderingId, collectionName) {
+    const renderingPermissions = PermissionsCache._getRenderingPermissions(renderingId);
+    if (renderingPermissions && renderingPermissions.data
+      && renderingPermissions.data[collectionName]) {
```
Comment (VincentMolinie) [inline src/services/permissions-cache.js:45]:
```suggestion
    if (renderingPermissions
      && renderingPermissions.data
      && renderingPermissions.data[collectionName]) {
```

File: src/services/permissions-cache.js [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+      ? PermissionsCache.transformPermissionsFromOldToNewFormat(permissions)
+      : null;
+
+    PermissionsCache.permissions.renderings = {
+      ...PermissionsCache.permissions.renderings,
+      [renderingId]: {
+        data: newFormatPermissions,
+        lastRetrieve: moment(),
+      },
+    };
```
Comment (VincentMolinie) [inline src/services/permissions-cache.js:104]: While doing a copy while you can just do 
```js
if (!PermissionsCache.permissions.renderings) {
  PermissionsCache.permissions.renderings = {}
}
PermissionsCache.permissions.renderings[renderingId] = { data: ... };
```

File: src/services/permissions-cache.js [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+      data: permissions.collections,
+      lastRetrieve: moment(),
+    };
+    PermissionsCache.permissions.renderings = {
+      ...PermissionsCache.permissions.renderings,
+      [renderingId]: {
+        data: permissions.renderings ? permissions.renderings[renderingId] : null,
+        lastRetrieve: moment(),
+      },
+    };
```
Comment (VincentMolinie) [inline src/services/permissions-cache.js:118]: Same Here

File: src/services/permissions-cache.js [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+    if (!lastRetrieve) {
+      return true;
+    }
+
+    const elapsedSeconds = currentTime.diff(lastRetrieve, 'seconds');
+    return elapsedSeconds >= PermissionsCache.expirationInSeconds;
+  }
+
+  static async retrievePermissions(environmentSecret, renderingId) {
```
Comment (VincentMolinie) [inline src/services/permissions-cache.js:162]: As the function to retrieve the permissions is here plus there is also all the function. to get the permissions I think this service should be called `PermissionsGetter`

File: src/services/permissions-checker.js [hunk 17]
```diff
@@ -304,7 +158,7 @@ class PermissionsChecker {
   async checkPermissions(collectionName, permissionName, permissionInfos) {
     // TODO IN NEXT PR: Distinguish collectionsPermissionExpired and scopePermissionExpired for
     // rolesACL format and retrieve only scope when needed.
-    if (this._isPermissionExpired(permissionName)) {
+    if (PermissionsCache.isPermissionExpired(this.renderingId, permissionName)) {
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:161]: Well the `isPermissionExpired` should actually be a private function. And here you should have only one function to get the permissions
```js
const permissions = await PermissionsGetter.getPermissionsForRendering(this.renderingId);
```

Comment (VincentMolinie) [reply src/services/permissions-checker.js:161]: This get permissions should return **ALWAYS** the same format and normally you should have no need to call `PermissionsGetter` anywhere else in this file 😉

File: src/services/permissions-checker.js [hunk 18]
```diff
@@ -241,7 +119,7 @@ class PermissionsChecker {
   }
 
   async _isAllowed(collectionName, permissionName, permissionInfos) {
-    const collectionsPermissions = PermissionsChecker.getCollectionsPermissions(this.renderingId);
+    const collectionsPermissions = PermissionsCache.getPermissions(this.renderingId);
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:122]: Like I said the permissions should be retrieved only once

File: src/services/permissions-checker.js [hunk 19]
```diff
@@ -259,7 +137,7 @@ class PermissionsChecker {
         return PermissionsChecker._isCollectionBrowseAllowed(
           collectionsPermissions[collectionName],
           permissionInfos,
-          this._getScopePermissions(collectionName),
+          PermissionsCache.getScopePermissions(this.renderingId, collectionName),
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:140]: Same here

File: src/services/permissions-checker.js [hunk 20]
```diff
@@ -269,32 +147,8 @@ class PermissionsChecker {
...
-      return true;
-    }
-
-    const elapsedSeconds = currentTime.diff(lastRetrieve, 'seconds');
-    return elapsedSeconds >= PermissionsChecker.expirationInSeconds;
-  }
-
   async _retrievePermissionsAndCheckAllowed(collectionName, permissionName, permissionInfos) {
-    await this._retrievePermissions();
+    await PermissionsCache.retrievePermissions(this.environmentSecret, this.renderingId);
```
Comment (VincentMolinie) [inline src/services/permissions-checker.js:151]: Not the job of this service. The `PermissionsGetter` should decide by itself if it should retrieve the permissions or not

File: test/services/permissions.test.js [hunk 21]
```diff
@@ -101,8 +90,8 @@ describe('services > permissions', () => {
                 },
               });
 
-            await new PermissionsChecker('envSecret', 1).checkPermissions('Users', 'browseEnabled', { userId: 1 })
-              .then(() => { expect(true).toStrictEqual(true); });
+            expect(() => new PermissionsChecker('envSecret', 1).checkPermissions('Users', 'browseEnabled', { userId: 1 }))
+              .not.toThrow();
```
Comment (VincentMolinie) [inline test/services/permissions.test.js:94]: I'm pretty sure this is not the right syntax 🤔 
```suggestion
const permissionsChecker = new PermissionsChecker('envSecret', 1);
await expect(permissionsChecker.checkPermissions('Users', 'browseEnabled', { userId: 1 })).not.toThrow();
```

File: test/services/permissions.test.js [hunk 22]
```diff
@@ -127,8 +116,8 @@ describe('services > permissions', () => {
                 },
               });
 
-            await new PermissionsChecker('envSecret', 2).checkPermissions('Users', 'browseEnabled', { userId: 1 })
-              .then(() => { expect(true).toStrictEqual(true); });
+            expect(() => new PermissionsChecker('envSecret', 2).checkPermissions('Users', 'browseEnabled', { userId: 1 }))
```
Comment (VincentMolinie) [inline test/services/permissions.test.js:119]: Same here and etc...

**Review by VincentMolinie — CHANGES_REQUESTED** [review]

File: src/services/permissions-checker.js (hunk 20 above)
Comment (adriguy) [reply src/services/permissions-checker.js]: Well actually no... because the permissions retrieval will depend on whether the cached permission is allowed or not!
So we'll need one service that will interfere with the other.
I left the retrievePermissions call in the permissions checker for now, let's discuss about it when possible.

File: src/services/permissions-checker.js (hunk 17 above)
Comment (adriguy) [reply src/services/permissions-checker.js]: Ok for the same format of the `PermissionsGetter.getPermissions`.
However, `isPermissionExpired` should not be a private function to me (see my comment below).

File: src/services/permissions-getter.js [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+
+  static getLastRetrieveTime(renderingId, permissionName) {
+    // In the case of rolesACL format and browseEnabled permission, the last retrieve to be taken
+    // into account is the one stored by rendering (because of the scope information).
+    if (PermissionsGetter.isRolesACLActivated && PermissionsGetter._getPermissionsInCollections() && permissionName !== 'browseEnabled') {
+      return PermissionsGetter._getPermissionsInCollections().lastRetrieve;
+    }
+    return PermissionsGetter._getPermissionsInRendering(renderingId)
+      ? PermissionsGetter._getPermissionsInRendering(renderingId).lastRetrieve
+      : null;
```
Comment (VincentMolinie) [inline src/services/permissions-getter.js:155]: You have an issue ere, but not a big one. I'll talk about it in your next PR

**Review by VincentMolinie — CHANGES_REQUESTED** [review]

File: src/services/permissions-getter.js (hunk 23 above)
Comment (adriguy) [reply src/services/permissions-getter.js:155]: This function will disappear in the next PR anyway 😉  But sure let's talk about it whenever you want!

Comment (adriguy) [reply src/services/permissions-getter.js:155]: Oh I think I know what you mean. It concerns the "browseEnabled" case when you can have retrieve only on renderings?
If it is, don't worry I take care of this in the next PR 👍 
In this PR, for rolesACL, the lastRetrieve are the same in collectionPermissions and in renderingPermissions.

**Review by VincentMolinie — APPROVED** [review]
Good Job !
